### PR TITLE
Mangler - reuse removed vars

### PR DESCRIPTION
**Pull request by boopathi**

### CODE REVIEW

File: packages/babel-plugin-minify-mangle-names/src/index.js [hunk 1]
```diff
@@ -17,6 +17,32 @@ module.exports = ({ types: t }) => {
...
+      }
+    }
+
+    hasReference(scope, name) {
+      return this.references.has(scope) && this.references.get(scope).has(name);
+    }
+
+    updateReference(scope, oldName, newName) {
+      if (!this.references.has(scope)) {
+        return;
```
Comment (boopathi) [inline packages/babel-plugin-minify-mangle-names/src/index.js]: this should throw ... new references aren't created when mangling.

**Comment by boopathi** [comment]
<details>
<summary>We can do better for this - </summary>

```js
function Foo() {
  var a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
  var A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
  var $, _;
  a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
  A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
  $, _;
  function Foo() {
    var a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
    var A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
    var $, _;
    a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
    $, _;
    function Foo() {
      var a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
      var A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
      var $, _;
      a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
      A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
      $, _;
    }
  }
}
```

</details>

**Comment by boopathi** [comment]
Applying this to jQuery - 

```
           raw     raw win gzip    gzip win parse time run    
babili     94.61kB 217%    32.71kB 158%     8ms        3952ms 
uglify     94.27kB 218%    32.78kB 158%     9ms        2427ms 
closure    94.23kB 218%    33.38kB 153%     9ms        10938ms
closure js 95.64kB 213%    33.78kB 150%     9ms        14664ms
```

File: packages/babel-plugin-minify-mangle-names/src/index.js [hunk 2]
```diff
@@ -17,6 +17,36 @@ module.exports = ({ types: t }) => {
...
+      const references = this.references.get(scope);
+      if (!references.has(oldName)) {
+        // already renamed
+        return;
+      }
+      references.delete(oldName);
+      references.add(newName);
+    }
+
+    hasBinding(scope, name) {
```
Comment (boopathi) [inline packages/babel-plugin-minify-mangle-names/src/index.js]: TODO: remove

File: packages/babel-plugin-minify-mangle-names/src/index.js [hunk 3]
```diff
@@ -101,11 +139,11 @@ module.exports = ({ types: t }) => {
           // Re-enable after enabling this feature
```
Comment (boopathi) [inline packages/babel-plugin-minify-mangle-names/src/index.js]: remove comments

File: packages/babel-plugin-minify-mangle-names/src/index.js [hunk 4]
```diff
@@ -136,15 +174,16 @@ module.exports = ({ types: t }) => {
...
             } while (
               !t.isValidIdentifier(next)
-              || hop.call(bindings, next)
+              || scope.hasBinding(next)
               || scope.hasGlobal(next)
-              || scope.hasReference(next)
+              || mangler.hasReference(scope, next)
             );
 
             // TODO:
```
Comment (boopathi) [inline packages/babel-plugin-minify-mangle-names/src/index.js]: remove comments

File: packages/babel-plugin-minify-mangle-names/src/index.js [hunk 5]
```diff
@@ -136,15 +174,16 @@ module.exports = ({ types: t }) => {
...
             // re-enable this - check above
-            // resetNext();
-            mangler.rename(scope, oldName, next);
+            resetNext();
+            mangler.rename(scope, binding, oldName, next);
+            mangler.updateReference(scope, oldName, next);
```
Comment (boopathi) [inline packages/babel-plugin-minify-mangle-names/src/index.js]: probably not necessary... have to double check

**Comment by kangax** [comment]
@boopathi if you found no issues, and our tests pass, I would merge this (after removing comments and all that). Then I'll build this locally and try it on our codebase.

**Comment by boopathi** [comment]
There are some bugs I'm noticing though .. Will update once I fix those ...

**Comment by boopathi** [comment]
Updated and it looks like its working fine ... Except for one bug - when run with other plugins and when other plugins modify scope information, even though mangler is run as a separate pass, it disturbs this scope information and what we track in references becomes invalid for certain cases. 

<details>
<summary>Example:</summary>

```js
function foo() {
  path.traverse({
    VariableDeclaration: function VariableDeclaration(varPath) {
      // somehow these two affect the causing of the error
      if (!varPath.isVariableDeclaration({ kind: "var" })) return;
      if (!isSameFunctionScope(varPath, path)) return;

      var _iteratorNormalCompletion5 = true;
      var _didIteratorError5 = false;
      var _iteratorError5 = undefined;

      try {
        x();
      } catch (err) {
         // errors here in this scope
        _didIteratorError5 = true;
      }
    }
  });
}
```
</details>

**Comment by boopathi** [comment]
weird that the tests pass on local, clearing cache on travis and re-running tests.

**Comment by kangax** [comment]
> Except for one bug - when run with other plugins and when other plugins modify scope information, even though mangler is run as a separate pass, it disturbs this scope information and what we track in references becomes invalid for certain cases.

So we're kind of back to square 1? :( This is the exact issues I ran into when integrating mangler back in the days.

**Comment by boopathi** [comment]
I fixed those things by creating them [on the fly](https://github.com/babel/babili/pull/243/commits/40a331cb504e648788ca0788b15dd1b9fb1cb946) ... and added some comments regarding the same ... Now it works

**Comment by kangax** [comment]
Ok, I'll merge this in and try it out in a bit

**Pull request merged**